platform/android/src/text/collator.cpp

```diff
@@ -0,0 +1,199 @@
+    return javaClass.New(env, constructor, language);
+}
+
+jni::Object<Locale> Locale::New(jni::JNIEnv& env, jni::String language, jni::String region) {
+    static auto constructor = javaClass.GetConstructor<jni::String, jni::String>(env);
+    return javaClass.New(env, constructor, language, region);
+}
+
+jni::Class<Locale> Locale::javaClass;
+
+} // namespace android
+
+namespace style {
+namespace expression {
+
+class Collator::Impl {
+public:
+    Impl(bool caseSensitive_, bool diacriticSensitive_, optional<std::string> locale_)
+        : caseSensitive(caseSensitive_)
+        , diacriticSensitive(diacriticSensitive_)
+        , env(android::AttachEnv())
+    {
+        LanguageTag languageTag = locale_ ? LanguageTag::fromBCP47(*locale_) : LanguageTag();
+        if (!languageTag.language) {
+            locale = android::Locale::getDefault(*env).NewGlobalRef(*env);
+        } else if (!languageTag.region) {
+            locale = android::Locale::New(*env,
+                                          jni::Make<jni::String>(*env, *(languageTag.language)))
+                    .NewGlobalRef(*env);
+        } else {
+            locale = android::Locale::New(*env,
+                                          jni::Make<jni::String>(*env, *(languageTag.language)),
+                                          jni::Make<jni::String>(*env, *(languageTag.region)))
+                    .NewGlobalRef(*env);
+        }
+        collator = android::Collator::getInstance(*env, *locale).NewGlobalRef(*env);;
+        if (!diacriticSensitive && !caseSensitive) {
+            android::Collator::setStrength(*env, *collator, 0 /*PRIMARY*/);
+        } else if (diacriticSensitive && !caseSensitive) {
+            android::Collator::setStrength(*env, *collator, 1 /*SECONDARY*/);
+        } else if (caseSensitive) {
+            // If we're case-sensitive and diacritic-sensitive, we use a case-sensitive collator
+            // and a fallback implementation of diacritic-insensitivity.
+            android::Collator::setStrength(*env, *collator, 2 /*TERTIARY*/);
+        }
+    }
+
+    bool operator==(const Impl& other) const {
+        return caseSensitive == other.caseSensitive &&
+                diacriticSensitive == other.diacriticSensitive &&
+                resolvedLocale() == other.resolvedLocale();
+    }
+
+    int compare(const std::string& lhs, const std::string& rhs) const {
+        bool useUnaccent = !diacriticSensitive && caseSensitive;
+        // java.text.Collator doesn't support a diacritic-insensitive/case-sensitive collation
+        // order, so we have to compromise here. We use Android's case-sensitive Collator
+        // against strings that have been "unaccented" using non-locale-aware nunicode logic.
+        // Because of the difference in locale-awareness, this means turning on case-sensitivity
+        // can _potentially_ change compare results for strings that don't actually have any case
+        // differences.
```

Comment:
Couldn't we use `java.text.Collator`'s `NO_DECOMPOSITION` flag instead of removing accents from the strings ourselves for case-sensitive and diacritic-insensitive collation?

Reply:
I don't _think_ that's what `NO_DECOMPOSITION` does -- I think using that flag actually kind of makes your collator _extra_ sensitive to diacritic changes, in that a "precomposed ü" will not be equal to "u + 'COMBINING DIAERESIS' (U+0308)", even though they'll display exactly the same way.

It's definitely possible there's a way to do the comparison we want entirely using `java.text.Collator`, I just haven't figured it out. The underlying ICU implementation (see http://userguide.icu-project.org/collation/architecture) has a `UCOL_SHIFTED` option that changes the meaning of primary/secondary/tertiary strength, and can be used for case-sensitive/diacritic-insensitve comparison. But I don't see a way to get at that from the java interface.

Reply:
Err, sorry I got confused, `UCOL_SHIFTED` is useful for ignoring punctuation, I was thinking of `UCOL_ON`:
```
UCOL_CASE_LEVEL == UCOL_ON and UCOL_STRENGTH == UCOL_PRIMARY
Used to ignore only the accent differences
```